include/net/net_pkt_filter.h

```diff
@@ -0,0 +1,364 @@
+
+/** @} */
+
+/**
+ * @defgroup npf_basic_cond Basic Filter Conditions
+ * @ingroup net_pkt_filter
+ * @{
+ */
+
+struct npf_test_iface {
+	struct npf_test test;
+	struct net_if *iface;
+};
+
+extern npf_test_fn_t npf_iface_match;
+extern npf_test_fn_t npf_iface_unmatch;
+
+/**
+ * @brief Define an "interface match" packet filter condition
+ *
+ * @param _name Name of the condition
+ * @param _iface Interface to match
+ */
+#define NPF_IFACE_MATCH(_name, _iface) \
+	struct npf_test_iface _name = { \
+		.iface = (_iface), \
+		.test.fn = npf_iface_match, \
+	}
+
+/**
+ * @brief Statically define an "interface unmatch" packet filter condition
+ *
+ * @param _name Name of the condition
+ * @param _iface Interface to exclude
+ */
+#define NPF_IFACE_UNMATCH(_name, _iface) \
+	struct npf_test_iface _name = { \
+		.iface = (_iface), \
+		.test.fn = npf_iface_unmatch, \
+	}
+
+struct npf_test_size_bounds {
+	struct npf_test test;
+	size_t min;
+	size_t max;
+};
+
+extern npf_test_fn_t npf_size_inbounds;
+
+/**
+ * @brief Statically define a "data minimum size" packet filter condition
+ *
+ * @param _name Name of the condition
+ * @param _size Lower bound of the packet's data size
+ */
+#define NPF_SIZE_MIN(_name, _size) \
+	struct npf_test_size_bounds _name = { \
+		.min = (_size), \
+		.max = SIZE_MAX, \
+		.test.fn = npf_size_inbounds, \
+	}
+
+/**
+ * @brief Statically define a "data maximum size" packet filter condition
+ *
+ * @param _name Name of the condition
+ * @param _size Higher bound of the packet's data size
+ */
+#define NPF_SIZE_MAX(_name, _size) \
+	struct npf_test_size_bounds _name = { \
+		.min = 0, \
+		.max = (_size), \
+		.test.fn = npf_size_inbounds, \
+	}
+
+/**
+ * @brief Statically define a "data bounded size" packet filter condition
+ *
+ * @param _name Name of the condition
+ * @param _min_size Lower bound of the packet's data size
+ * @param _max_size Higher bound of the packet's data size
+ */
+#define NPF_SIZE_BOUNDS(_name, _min_size, _max_size) \
+	struct npf_test_size_bounds _name = { \
+		.min = (_min_size), \
+		.max = (_max_size), \
+		.test.fn = npf_size_inbounds, \
+	}
+
+/** @} */
+
+/**
+ * @defgroup npf_eth_cond Ethernet Filter Conditions
+ * @ingroup net_pkt_filter
+ * @{
+ */
+
+struct npf_test_eth_addr {
+	struct npf_test test;
+	unsigned int nb_addresses;
+	struct net_eth_addr *addresses;
+};
+
+extern npf_test_fn_t npf_eth_src_addr_match;
+extern npf_test_fn_t npf_eth_src_addr_unmatch;
+extern npf_test_fn_t npf_eth_dst_addr_match;
+extern npf_test_fn_t npf_eth_dst_addr_unmatch;
+
+/**
+ * @brief Statically define a "source address match" packet filter condition
+ *
+ * This tests if the packet source address matches any of the Ethernet
+ * addresses contained in the provided set.
+ *
+ * @param _name Name of the condition
+ * @param _addr_array Array of <tt>struct net_eth_addr</tt> items to test against
+ */
+#define NPF_ETH_SRC_ADDR_MATCH(_name, _addr_array) \
```

Comment:
I am not a big fan of adding Ethernet specific stuff to this generic API, would it be possible to make this more generic so that other L2 protocols could be matched? This way we could avoid creating new APIs to other network technologies.